kernel/sched.c

```diff
@@ -959,6 +962,12 @@ void *z_get_next_switch_handle(void *interrupted)
 			 * will not return into it.
 			 */
 			if (z_is_thread_queued(old_thread)) {
+#ifdef CONFIG_SCHED_IPI_CASCADE
+				if ((new_thread->base.cpu_mask != -1) &&
+				    (old_thread->base.cpu_mask & BIT(cpu_id))) {
```

Comment:
That test for old_thread is wrong, I think?  You want `old_thread->base.cpu_mask != BIT(cpu_id)`, because you know you can skip the IPI if the outgoing thread isn't capable of running anywhere else.  The test as written here should be true by definition for a preempted thread, since it was, after all, running on this CPU (though I guess in an edge case it's possible that it was re-pinned in the ISR and that's why it's being preempted).

Also you probably want to add a case for `!IS_ENABLED(CONFIG_CPU_MASK_PIN_ONLY)`, which guarantees that the cascade isn't needed because nothing is able to migrate anyway.

Reply:
You're right .. that check is wrong. (I had meant to do `~BIT(cpu_id)`, but I like your simplification better.)

Unless I am mistaken, the `!IS_ENABLED(CONFIG_CPU_MASK_PIN_ONLY)` is not required as that ought to be handled by the Kconfig.smp as SCHED_IPI_CASCADE is dependent upon SCHED_CPU_MASK_PIN_ONLY not being set.